File: articles/calculus-for-systems-modeling/articles/systems-of-differential-equations/fortran/coupled_system_audit.f90
program coupled_system_audit
  implicit none
  real(8) :: prey, predator, alpha, beta, delta, gamma, dt, t, prey_rate, predator_rate
  integer :: n, steps
  prey = 40.0d0; predator = 9.0d0; alpha = 0.7d0; beta = 0.05d0; delta = 0.02d0; gamma = 0.5d0; dt = 0.01d0; steps = 2000
  print '(A)', 'scenario time prey predator prey_rate predator_rate alpha beta delta gamma method'
  do n = 0, steps
    t = n*dt
    prey_rate = alpha*prey - beta*prey*predator
    predator_rate = delta*prey*predator - gamma*predator
    print '(A,9F12.6,1X,A)', 'predator_prey_coupled_system', t, prey, predator, prey_rate, predator_rate, alpha, beta, delta, gamma, 'explicit_euler'
    prey = max(0.0d0, prey + dt*prey_rate)
    predator = max(0.0d0, predator + dt*predator_rate)
  end do
end program

File: articles/calculus-for-systems-modeling/articles/systems-of-differential-equations/cpp/coupled_system_audit.cpp
#include <algorithm>
#include <iomanip>
#include <iostream>
#include <utility>

std::pair<double,double> rates(double prey, double predator, double alpha, double beta, double delta, double gamma){
  return {alpha*prey - beta*prey*predator, delta*prey*predator - gamma*predator};
}

int main(){
  double prey = 40.0, predator = 9.0, alpha = 0.7, beta = 0.05, delta = 0.02, gamma = 0.5, dt = 0.01;
  int steps = 2000;
  std::cout << std::fixed << std::setprecision(6);
  std::cout << "scenario,time,prey,predator,prey_rate,predator_rate,alpha,beta,delta,gamma,method,warning\n";
  for(int n=0; n<=steps; n++){
    double t = n*dt;
    auto [prey_rate, predator_rate] = rates(prey, predator, alpha, beta, delta, gamma);
    std::cout << "predator_prey_coupled_system," << t << "," << prey << "," << predator << "," << prey_rate << "," << predator_rate << ","
              << alpha << "," << beta << "," << delta << "," << gamma << ",explicit_euler,Predator-prey terms are illustrative and assume continuous well-mixed interaction.\n";
    prey = std::max(0.0, prey + dt*prey_rate);
    predator = std::max(0.0, predator + dt*predator_rate);
  }
}
